### PR TITLE
Web import parsers

### PR DESCRIPTION
**Pull request by FriedrichFroebel**
This is my attempt to update the website import parsers for the latest website versions.

Some notes:

* I followed the naming scheme used by the existing schema.org parser.
* I added some tests for websites which did not have one.
* I added the option to either use the remote or the local HTML for the tests.
* I removed an old file which did not seem to be useful at all.
* I left the `about_dot_com_plugin` and `cooksillustrated_plugin` as-is, although I am not sure if they are relevant. I could not find a webpage for the first one, while the second one requires a login.
* For parsing the schema.org stuff, I introduced the ` scrape-schema-recipe` dependency.
* I am not sure whether this breaks the communication between the plugins and the application itself, but at least the terminal-based stuff from the test works for me. We will probably have to adapt some interface stuff later.

### CODE REVIEW

File: gourmet/gglobals.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
         # http://stackoverflow.com/questions/2608200/problems-with-umlauts-in-python-appdata-environvent-variable
         # We might drop this workaround with Python 3 (all strings are unicode)
         # and/or GTK+ 3 (use Glib.get_home_dir()).
-        APPDATA = windows.getenv('APPDATA').decode('utf-8')
+        APPDATA = windows.getenv('APPDATA')
```
Comment (maweki) [inline gourmet/gglobals.py]: this shouldn't be part of the PR

File: gourmet/Undo.py [hunk 2]
```diff
@@ -1,4 +1,6 @@
 import difflib, re
+import gi
+gi.require_version("Gtk", "3.0")
```
Comment (maweki) [inline gourmet/Undo.py]: this shouldn't be part of the PR

File: requirements.txt [hunk 3]
```diff
@@ -7,7 +7,8 @@ Pillow
...
 pyenchant
 pygobject
-# pygtkspelaplcheck
+# pygtkspellcheck
 reportlab
+scrape-schema-recipe
 selenium
-sqlalchemy
\ No newline at end of file
+sqlalchemy
```
Comment (maweki) [inline requirements.txt]: Can you tell us more about these changes? Why is sqlalchemy now a dependency?

**Comment by maweki** [comment]
The size of the PR borders on unreviewable, but I try to soldier through. Might need some help from @FriedrichFroebel  ;)

File: gourmet/gglobals.py (hunk 1 above)
Comment (FriedrichFroebel) [reply gourmet/gglobals.py]: Otherwise I could not run the tests.

File: requirements.txt (hunk 3 above)
Comment (FriedrichFroebel) [reply requirements.txt]: `sqlalchemy` has already been a dependency before. I just fixed the naming for `pygtkspellcheck` and added the `scrape-schema-recipe` requirement.

File: gourmet/gglobals.py (hunk 1 above)
Comment (maweki) [reply gourmet/gglobals.py]: That might be, but you don't need to commit all your changes. That should be another PR.

**Comment by FriedrichFroebel** [comment]
Most of the changes actually should be related to just updating the local HTML files for the different websites - the parsing stuff is much less in this PR.

File: gourmet/plugins/import_export/website_import_plugins/foodnetwork_plugin.py [hunk 4]
```diff
@@ -6,42 +6,49 @@ class FoodNetworkPlugin(PluginPlugin):
     target_pluggable = 'webimport_plugin'
 
     def test_url(self, url, data):
-        if 'foodnetwork.com' in url:
+        if 'foodnetwork.co.uk' in url:
             return 5
         return 0
```
Comment (maweki) [inline gourmet/plugins/import_export/website_import_plugins/foodnetwork_plugin.py]: I know, you didn't introduce them, but is there a reason for the strange return code and would it be better to introduce an enum for this?

**Comment by cydanil** [comment]
It looks like it's been quite the ride!
Because there are some many changes, expect that there will be a few comments :)

Thanks for taking the time to go through it

File: gourmet/tests/test_allrecipes_com_plugin.py [hunk 5]
```diff
@@ -1,29 +1,37 @@
...
         filename = os.path.join(os.path.dirname(__file__),
-                                'recipe_files',
-                                (os.path.splitext(os.path.basename(__file__))[0])[5:-7] + ".html")
-        with open(filename) as f:
+                                "recipe_files",
+                                "allrecipes_com.html")
+        with open(filename, encoding="utf8") as f:
             data = f.read()
-        # print(data)
         return data
```
Comment (maweki) [inline gourmet/tests/test_allrecipes_com_plugin.py]: This function either returns `data` or sets `self.text`. Can we have that consistent? If it always returns `data`, we can have it as a `staticmethod`, which would be nice.

File: gourmet/plugins/import_export/website_import_plugins/cooksillustrated.py [hunk 6]
```diff
@@ -1,64 +0,0 @@
-from selenium import webdriver
```
Comment (cydanil) [inline gourmet/plugins/import_export/website_import_plugins/cooksillustrated.py:1]: Could you quickly say why this file is deleted?

File: gourmet/tests/test_allrecipes_com_plugin.py [hunk 7]
```diff
@@ -1,29 +1,37 @@
...
 
     def setUp(self):
-        self.text = self._read_html()
+        self.text = self._read_html(False)
```
Comment (maweki) [inline gourmet/tests/test_allrecipes_com_plugin.py]: If `download` is True, then None is returned and self.text is empty, even though `_read_html` had it set.

File: gourmet/tests/test_epicurious_plugin.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            with urllib.request.urlopen(self.url) as response:
+                self.text = response.read().decode("utf8")
+            return
+
+        filename = os.path.join(os.path.dirname(__file__),
+                                "recipe_files",
+                                "epicurious_com.html")
+        with open(filename, encoding="utf8") as f:
+            data = f.read()
+        return data
```
Comment (maweki) [inline gourmet/tests/test_epicurious_plugin.py]: Same issue: inconsistent return types. Please convert to `staticmethod` and always return the data

File: gourmet/tests/test_foodnetwork_plugin.py [hunk 9]
```diff
@@ -5,32 +5,39 @@
...
 class TestFoodnetworkPlugin(unittest.TestCase):
 
-    url = "http://www.foodnetwork.com/recipes/ask-aida/pan-roasted-chicken-with-oranges-and-rosemary-recipe/index.html"
+    url = "https://foodnetwork.co.uk/recipes/pan-roasted-chicken-thighs-grapes-and-olives/"
+
+    def _read_html(self, download=True):
+        if download:
+            with urllib.request.urlopen(self.url) as response:
+                self.text = response.read().decode("utf8")
+            return
```
Comment (maweki) [inline gourmet/tests/test_foodnetwork_plugin.py]: Here as well

File: gourmet/tests/test_foodnetwork_plugin.py [hunk 10]
```diff
@@ -49,20 +56,20 @@ def test_parse(self):
...
+        self.assertEqual(len(ingredients), 11)
+        self.assertEqual(preptime, "15 mins")
+        self.assertTrue(cooktime, "25 mins")
+        self.assertEqual(yields, "4")
 
-        self.assertFalse('You must be logged in to review this recipe.' in instructions)
+        self.assertEqual(name, "Pan Roasted Chicken Thighs with Grapes and Olives")
 
+        self.assertTrue(
+            "Place the skillet underneath the broiler to crisp the chicken skin, about 2 minutes. Watch carefully to avoid burning." in instructions)
```
Comment (maweki) [inline gourmet/tests/test_foodnetwork_plugin.py]: Please replace with https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIn for cooler error messages

File: gourmet/tests/test_epicurious_plugin.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        # Check results
+        self.assertEqual(len(ingredients), 3)
+        self.assertEqual(yields, "Makes about 3 1/2 cups")
+        self.assertEqual(cooktime, "4 hours (includes churning, freezing, and softening time)")
+
+        self.assertEqual(name, "Kiwi-Lime Sorbet")
+        self.assertTrue("A perfect ending to any Asian meal." in modifications)
+        self.assertTrue("Puree all ingredients in processor." in instructions)
+        self.assertTrue(
+            "Process in ice cream maker according to manufacturer's instructions. Transfer to container, cover, and freeze until solid, at least 3 hours. (Can be made 2 days ahead. Let stand at room temperature 30 minutes before serving.)" in instructions)
```
Comment (maweki) [inline gourmet/tests/test_epicurious_plugin.py]: Please replace with https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIn for cooler error messages

File: gourmet/tests/test_ica_se_plugin.py [hunk 12]
```diff
@@ -1,75 +1,77 @@
...
         filename = os.path.join(os.path.dirname(__file__),
-                                'recipe_files',
-                                (os.path.splitext(os.path.basename(__file__))[0])[5:-7]+".html")
-        with open(filename, encoding="utf8") as infile:
-            data = infile.read()
+                                "recipe_files",
+                                "ica_se.html")
+        with open(filename, encoding="utf8") as f:
+            data = f.read()
         return data
```
Comment (maweki) [inline gourmet/tests/test_ica_se_plugin.py]: Inconsistent return

File: gourmet/tests/test_ica_se_plugin.py [hunk 13]
```diff
@@ -1,75 +1,77 @@
...
 
-        print(type(instructions))
-        self.assertTrue('Dela varje kycklinglårfilé i 2 bitar.' in instructions,
-                instructions)
-        self.assertTrue('Ta upp dem och skölj ur pannan.' in instructions)
-        self.assertTrue('Fri från gluten, laktos, mjölkprotein och ägg.' in instructions)
+        self.assertTrue(
+            "Dela varje kycklinglårfilé i 2 bitar. Blanda chilipulver och soja i en skål. Lägg ner kycklingen och blanda runt. Låt stå i kylen 30 minuter." in instructions)
+        self.assertTrue(
+            "Rätten passar bra till en buffé. Portionerna är beräknade för att passa buffébordet." in modifications)
```
Comment (maweki) [inline gourmet/tests/test_ica_se_plugin.py]: assertIn

File: gourmet/tests/test_nytimes_plugin.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            with urllib.request.urlopen(self.url) as response:
+                self.text = response.read().decode("utf8")
+            return
+
+        filename = os.path.join(os.path.dirname(__file__),
+                                "recipe_files",
+                                "nytimes.html")
+        with open(filename, encoding="utf8") as f:
+            data = f.read()
+        return data
```
Comment (maweki) [inline gourmet/tests/test_nytimes_plugin.py]: same issue

File: gourmet/tests/test_nytimes_plugin.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        # Check results
+        self.assertEqual(len(ingredients), 14)
+        self.assertEqual(cooktime, "1 h")
+        self.assertEqual(yields, "4 servings")
+
+        self.assertEqual(name, "Egg Curry")
+        self.assertEqual(cuisine, "indian")
+
+        self.assertTrue(
+            "Add the tomatoes, salt and 1 cup water. Cook, stirring occasionally, until the mixture thickens and the fat rises to the top, about 15 minutes. Stir in the garam masala and lower the heat. If the sauce isn’t runny, stir in 1/2 cup water." in instructions)
```
Comment (maweki) [inline gourmet/tests/test_nytimes_plugin.py]: same issue

**Review by maweki — CHANGES_REQUESTED** [review]
Overall this looks good

* Please remove the two files with unrelated changes from the PR
* Please fix the `_get_html` methods to be `staticmethod` and always return the html data
* Please replace `assertTrue(a in b)` with `assertIn(a, b)`

Can you say something about the `return 5` in the scaffold? Is this some enum value we might want to extract?

File: gourmet/plugins/import_export/website_import_plugins/foodnetwork_plugin.py (hunk 4 above)
Comment (FriedrichFroebel) [reply gourmet/plugins/import_export/website_import_plugins/foodnetwork_plugin.py]: I have no clue about it. There seem to be 3 different values: `0` (no match), `5` (match) and `4` (some unknown return code used by `cooksillustrated_plugin`). I will add an own file with a corresponding enumeration class.

File: gourmet/tests/test_allrecipes_com_plugin.py (hunk 5 above)
Comment (FriedrichFroebel) [reply gourmet/tests/test_allrecipes_com_plugin.py]: This is some leftover from an old version. I will fix this.

File: gourmet/tests/test_epicurious_plugin.py (hunk 11 above)
Comment (FriedrichFroebel) [reply gourmet/tests/test_epicurious_plugin.py]: Perfectly makes sense - will replace it.

**Review by cydanil — COMMENTED** [review]
There are a few places where I'd want to nitpick, like "new blank lines, blablabla". It's no big deal here, though.

Overall, you did a fantastic job at fixing the importers. However, I would suggest in the future to split the PR into a number of smaller ones, as it's a bit hard to jump between an importer and its tests. For instance here, I think making a PR per website importer would be clearer :)

Perhaps in another PR, all the importer tests and their data could be moved to a subfolder, like `gourmet/tests/importers`? Maybe someone else has an input on this?

File: gourmet/tests/test_allrecipes_com_plugin.py (hunk 7 above)
Comment (FriedrichFroebel) [reply gourmet/tests/test_allrecipes_com_plugin.py]: Oops, will fix it.

Comment (maweki) [reply gourmet/tests/test_allrecipes_com_plugin.py]: This issue will be gone, once `_read_html` is a `staticmethod`

File: gourmet/plugins/import_export/website_import_plugins/cooksillustrated.py (hunk 6 above)
Comment (FriedrichFroebel) [reply gourmet/plugins/import_export/website_import_plugins/cooksillustrated.py:1]: This seemed to be an unused file used for testing some stuff in the past. The real plugin is located in `cooksillustrated_plugin.py` and does not use this file, so I assumed it should be safe to delete it.

**Review by maweki — APPROVED** [review]
I am basically fine with this. But maybe a second pair of eyes would be good.

**Comment by cydanil** [comment]
Yes! Thanks :)

**Review by cydanil — APPROVED** [review]
